amivapi/studydocs/rating.py

```diff
@@ -0,0 +1,81 @@
+# -*- coding: utf-8 -*-
+#
+# license: AGPLv3, see LICENSE for details. In addition we strongly encourage
+#          you to buy us beer if we meet and you like the software.
+"""Studydoc rating computation.
+
+For the rating, we use the lower bound of a confidence interval around
+the 'naive' average vote, which accounts for the number of votes.
+
+The rating for a study document is updated each time a study doc rating is
+for the respective document is POSTed or PATCHed. The value is then written
+to the database to allow sorting of study documents by rating.
+"""
+
+from flask import current_app
+from math import sqrt
+
+from amivapi.utils import get_id
+
+
+def compute_rating(upvotes, downvotes, z=1.28):
+    """Compute the rating.
+
+    Concretely, the lower bound of the wilson confidence interval is returned,
+    which takes the number of votes into account. [1]
+
+    We use z = 1.28 by default, which corresponds to a 80% confidence interval
+    (As there are only a few votes, we do not want to be overly cautious).
+
+    [1]: http://www.evanmiller.org/how-not-to-sort-by-average-rating.html
+    """
+    total = upvotes + downvotes
+    if not total:
+        return
+
+    p = upvotes / total
+    bound = ((p + (z**2)/(2*total) -
```

Comment:
not sure if python optimizes this somehow, but often z*z is vastly faster than z**2, as it uses a simple multiplication instead of a long algorithm.

Reply:
Not sure either, but I do not think it matters here. The algorithm runs once per vote, so the performance difference will be negligible.